### DISCUSSION TITLE
Visualize Prometheus metrics with Grafana

### DISCUSSION THREAD

**Discussion by s5uishida** [herlesupreeth/docker_open5gs#132] [General]
Hi @herlesupreeth 

Thank you very much for supporting Prometheus metrics.

I installed Grafana-OSS from DockerHub and was able to view the metrics.

**Comment by myonlystarWang** [comment]
Hi  @s5uishida 
    Did you follow this guidance completely? [(https://open5gs.org/open5gs/docs/tutorial/04-metrics-prometheus/)](url)

**Reply by s5uishida** [reply]
docker_open5gs contains the necessary settings, so I think you can use it without referring to that tutorial.

**Reply by myonlystarWang** [reply]
I got it. Thank you very much!

**Comment by herlesupreeth** [comment]
@s5uishida thank you for trying it out. Integrating with Grafana was on my todo list next :). Will probably ask you if I face any issues integrating prometheus with it.

**Reply by herlesupreeth** [reply]
@s5uishida btw, you can see graph in prometheus itself. Did you see any benefit in using Graphana vs Prometheus for visualizing graph?

**Reply by s5uishida** [reply]
Yes (probably).

I think it would be convenient to use it with Grafana to flexibly layout various panels and display them on the dashboard.
I don't know how to flexibly customize dashboards with Prometheus ;-)

**Reply by s5uishida** [reply]
> Will probably ask you if I face any issues integrating prometheus with it.

When access the Prometheus dashboard from the browser of the PC on the network outside DockerHost, you will not be able to see metrics for each `job_name` (`172.22.0.x/24` on Docker).

However, when connect to `http://DOCKER_HOST_IP:9090/metrics`, you will be able to view all `job_name`'s (NFs) metrics in a browser.

In the case of Grafana, the metrics on the dashboard on the browser will be referenced via the Grafana server, so I think Grafana and Prometheus will communicate within `172.22.0.0/24`.

So there is no need to expose `ports` to the network outside of DockerHost, and the network settings in the current `docker-compose.yaml `of docker_open5gs are fine.

You can view all NFs metrics from Grafana's Metrics browser (list box). If multiple `job_names` have the same metric, each `job_name` is displayed.

For your reference.

**Comment by s5uishida** [comment]
Hi @herlesupreeth 

As a trial, with reference to your `docker_open5gs/metrics`, I confirmed the operation of visualization of metrics with Grafana as follows.
(as a example)

Access URL: `http://DOCKER_HOST_IP:3001/`

docker_open5gs/.env
```
--- .env.orig   2022-11-24 09:48:11.000000000 +0900
+++ .env        2022-11-24 11:45:29.448965227 +0900
@@ -116,3 +116,6 @@
 
 # METRICS
 METRICS_IP=172.22.0.36
+
+# Grafana
+GRAFANA_IP=172.22.0.37
```
docker_open5gs/docker-compose.yaml
```
--- docker-compose.yaml.orig    2022-11-24 09:48:11.000000000 +0900
+++ docker-compose.yaml 2022-11-24 11:46:42.013296671 +0900
@@ -658,6 +658,23 @@
     networks:
       default:
         ipv4_address: ${METRICS_IP}
+  grafana:
+    build: ./grafana
+    image: docker_grafana
+    container_name: grafana
+    env_file:
+      - .env
+    volumes:
+      - ./grafana:/mnt/grafana
+      - /etc/timezone:/etc/timezone:ro
+      - /etc/localtime:/etc/localtime:ro
+    expose:
+      - "3000/tcp"
+    ports:
+      - "3001:3000/tcp"
+    networks:
+      default:
+        ipv4_address: ${GRAFANA_IP}
 networks:
   default:
     ipam:
```
docker_open5gs/grafana/Dockerfile
```
FROM ubuntu:focal

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install -y wget && \
    wget -q https://dl.grafana.com/oss/release/grafana-9.2.6.linux-amd64.tar.gz && \
    tar xvfz grafana-9.2.6.linux-amd64.tar.gz

WORKDIR grafana-9.2.6

CMD /mnt/grafana/grafana_init.sh
```
docker_open5gs/grafana/grafana_init.sh
```
#!/bin/bash

bin/grafana-server
```

**Reply by herlesupreeth** [reply]
Thank you. Will incorporate shortly. Currently busy with day-to-day job :sweat_smile: so will probably work on this over weekend.

**Comment by s5uishida** [comment]
A little information.

PCF and UPF now support Prometheus. AMF and SMF metrics have also been added a little.
For the list of metrics, please refer to the updated following.

https://github.com/s5uishida/open5gs_5gc_ueransim_metrics_sample_config

**Reply by herlesupreeth** [reply]
Ah nice!! Thanks for letting me know. Will add them.
